File: scripts/07_cell_cell_communication.R
"""
Worker Script: 07_cell_cell_communication.R
Project: Colorectal Cancer (CRC) Single-Cell Atlas
Stage: Per-Condition Signaling Inference & Pattern Discovery

This script iterates through condition-specific h5ad files (e.g., Normal, Tumor),
performs ligand-receptor interaction inference using CellChat, and executes
manifold learning to identify global communication patterns.

Input: Condition-specific .h5ad files
Output: .rds objects for each condition and diagnostic figures.

Author: Loai Eletr
"""

library(jsonlite)
library(reticulate)

# --- 1. ENVIRONMENT VALIDATION ---
# CellChat uses Python's UMAP for manifold learning (Pattern Analysis).
if (!py_module_available("umap")) {
  stop("❌ UMAP not found. Ensure your Python environment is correctly linked to R via reticulate.")
}

# --- 2. ARGUMENT PARSING ---
# These arguments are passed from the 07_cell_cell_communication.py orchestrator.
args <- commandArgs(trailingOnly = TRUE)
project_dir  <- args[1] # Root project directory
input_dir    <- args[2] # Directory containing split h5ad files
rds_dir      <- args[3] # Destination for processed R objects
fig_dir      <- args[4] # Destination for individual signaling plots
species      <- args[5] # 'human' or 'mouse'
min_cells    <- as.numeric(args[6]) # Threshold for cell group inclusion
k_json_path  <- args[7] # Path to JSON containing optimized K values

# --- 3. RESOURCE INITIALIZATION ---
source(file.path(project_dir, "src/cellchats.R")) # Custom wrapper functions
k_map <- fromJSON(k_json_path)                    # Load k_out and k_in per sample

# Create output directories if they don't exist
dir.create(rds_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(fig_dir, showWarnings = FALSE, recursive = TRUE)

# Discover the data files to be processed
files <- list.files(input_dir, pattern = "*.h5ad", full.names = TRUE)

# --- 4. MAIN PROCESSING LOOP ---
for (f in files) {
  name <- tools::file_path_sans_ext(basename(f))

  # Retrieve optimized K values from JSON, fallback to 3 if not found
  k_out <- if (!is.null(k_map[[name]])) k_map[[name]]$k_out else 3
  k_in  <- if (!is.null(k_map[[name]])) k_map[[name]]$k_in else 3

  cat("\n", rep("=", 50), "\n", sep="")
  cat("🚀 PROCESSING SAMPLE: ", name, "\n")
  cat(rep("=", 50), "\n", sep="")

  # --- STEP 1: INFERENCE ---
  # Initialize CellChat object and set the L-R database (e.g., CellChatDB.human)
  obj <- create_cc_obj(f)
  setup <- set_cc_db(obj, species)

  # Calculate communication probabilities and filter out rare cell groups
  obj <- run_cc_inference(setup$obj, setup$ppi, min_cells = min_cells)

  # --- STEP 2: DIAGNOSTIC K-SELECTION ---
  # Generates plots to verify if the number of patterns (k_out/k_in) is appropriate
  cat("[*] Generating K-selection diagnostics...\n")
  run_cc_k_selection(obj, name = name, plots_dir = file.path(fig_dir, name, "k_diagnostics"))

  # --- STEP 3: SYSTEMS ANALYSIS (PATTERNS) ---
  # Identifies which cell types are 'outgoing' (secretors) and 'incoming' (receivers)
  # across coordinated signaling pathways.
  cat("[*] Identifying communication patterns (k_out=", k_out, ", k_in=", k_in, ")...\n", sep="")
  obj <- run_cc_systems_analysis(
    obj, k_out = k_out, k_in = k_in, name = name,
    plots_dir = file.path(fig_dir, name, "individual_patterns")
  )

  # --- STEP 4: DATA PERSISTENCE ---
  # Save as RDS for use in the Stage 08 Differential Analysis
  saveRDS(obj, file = file.path(rds_dir, paste0(name, "_cellchat.rds")))
  cat("✅ Saved: ", paste0(name, "_cellchat.rds"), "\n")
}

cat("\n🎉 Individual processing complete. Ready for differential analysis.\n")

File: scripts/07_cell_cell_communication.py
"""
Orchestration Script: 07_cell_cell_communication
Project: Colorectal Cancer (CRC) Single-Cell Atlas
Stage: Cell-Cell Communication (CCC) Analysis

This script serves as the bridge between the Python/Scanpy workflow and
the R/CellChat ecosystem. It handles data partitioning by biological
condition (Normal vs. Tumor), environment preparation, and the execution
of the R-based signaling inference engine.

Author: Loai Eletr
"""

import subprocess
import config
import time
from config import CellChatConfig
from src.utils import save_adata, load_r_environment

# ==============================================================================
# MAIN EXECUTION BLOCK
# ==============================================================================

if __name__ == "__main__":
    print(f"\n{'=' * 60}")
    print(f"🧬 STARTING CELL-CELL COMMUNICATION ANALYSIS")
    print(f"{'=' * 60}")
    start_time = time.time()

    # --- 1. PREPARATION & DATA SPLITTING ---
    # Load R paths and ensure the environment is ready for subprocess execution.
    print(f"[*] Initializing R environment at: {config.R_HOME}")
    load_r_environment(config.R_HOME)

    print("[*] Preparing directories and K-selection configuration...")
    custom_env = CellChatConfig.prepare_environment()

    # Partitioning: CellChat performs best when comparing separate objects
    # (e.g., one for Normal, one for Primary Tumor).
    print("[*] Splitting integrated AnnData into condition-specific files...")
    CellChatConfig.split_data_by_condition(save_adata)
    print(f"    --> Inputs saved to: {CellChatConfig.CELLCHAT_IO['input_dir']}")

    # --- 2. R-SUBSYSTEM INVOCATION ---
    # Launching the R script provided in previous steps.
    # We pass all configuration parameters (Species, Min Cells, IO paths)
    # as command-line arguments to maintain a "Single Source of Truth".
    print(f"\n[!] Launching Rscript: scripts/08_cell_cell_communication.R")
    print(f"    Species: {CellChatConfig.CELLCHAT_PARAMS['species']}")
    print(f"    Min Cells: {CellChatConfig.CELLCHAT_PARAMS['min_cells']}")
    print(f"{'-' * 60}")

    try:
        # The core hand-off to the R CellChat implementation
        subprocess.run([
            "Rscript",
            f"{config.BASE_DIR}/scripts/07_cell_cell_communication.R",
            str(config.BASE_DIR),
            str(CellChatConfig.CELLCHAT_IO["input_dir"]),
            str(CellChatConfig.CELLCHAT_IO["individual_cellchat_dir"]),
            str(CellChatConfig.CELLCHAT_IO["fig_dir"]),
            CellChatConfig.CELLCHAT_PARAMS["species"],
            str(CellChatConfig.CELLCHAT_PARAMS["min_cells"]),
            str(CellChatConfig.CELLCHAT_IO["k_json"])
        ], env=custom_env, check=True)

        # --- 3. TELEMETRY & RESULTS ---
        elapsed = (time.time() - start_time) / 60

        print(f"{'-' * 60}")
        print(f"✅ ANALYSIS COMPLETE in {elapsed:.2f} minutes")
        print(f"📂 RDS Objects: {CellChatConfig.CELLCHAT_IO['individual_cellchat_dir']}")
        print(f"📊 Figures:     {CellChatConfig.CELLCHAT_IO['fig_dir']}")
        print(f"{'=' * 60}\n")

    except subprocess.CalledProcessError as e:
        # ERROR HANDLING: Captures issues in the R script (e.g., missing libraries,
        # lack of signaling in a specific subset) and reports it to the Python log.
        print(f"\n❌ ERROR: Rscript failed with exit code {e.returncode}")
        print(f"Check the R logs above for specific CellChat errors.")
        print(f"{'=' * 60}\n")